tests/unit/test_endpoints_v2.py

```diff
@@ -0,0 +1,555 @@
+# Copyright 2012-2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+import json
+import logging
+import os
+
+import pytest
+
+from botocore.endpoints_v2 import (
+    EndpointInputParametersError,
+    EndpointProvider,
+    EndpointResolutionError,
+    EndpointRule,
+    ErrorRule,
+    RuleCreator,
+    RuleSetConfigurationError,
+    RuleSetEndpoint,
+    TreeRule,
+)
+from botocore.loaders import Loader
+
+
+def region_template():
+    return '{Region}'
+
+
+def region_ref():
+    return {'ref': 'Region'}
+
+
+def bucket_arn_ref():
+    return {'ref': 'bucketArn'}
+
+
+def parse_arn_func():
+    return {
+        'fn': 'aws.parseArn',
+        'argv': [{'ref': 'Bucket'}],
+        'assign': 'bucketArn',
+    }
+
+
+def string_equals_func():
+    return {
+        'fn': 'stringEquals',
+        'argv': [
+            {
+                'fn': 'getAttr',
+                'argv': [bucket_arn_ref(), 'region'],
+                'assign': 'bucketRegion',
+            },
+            '',
+        ],
+    }
+
+
+def url_template():
+    return f'https://{region_template()}.myGreatService.PartitionResults#dnsSuffix'
+
+
+def endpoint_dict():
+    return {
+        'url': url_template(),
+        'properties': {
+            'authSchemes': [
+                {
+                    "signingName": "s3",
+                    "signingScope": region_template(),
+                    "name": "s3v4",
+                }
+            ],
+        },
+        'headers': {
+            'x-amz-region-set': [
+                # more than one value in a given header is not allowed
+                region_ref(),
+                {
+                    'fn': 'getAttr',
+                    'argv': [bucket_arn_ref(), 'region'],
+                },
+                'us-east-2',
+            ],
+        },
+    }
+
+
+@pytest.fixture
+def partitions():
+    loader = Loader()
+    return loader.load_data('partitions')
+
+
+@pytest.fixture
+def ruleset_dict():
+    path = os.path.join(
+        os.path.dirname(__file__), 'data', 'endpoints', 'rules', 'ruleset.json'
+    )
+    with open(path) as f:
+        data = f.read()
+        return json.loads(data)
```

Comment:
Is this module scope? If we're not sure, let's do a quick once over on these fixtures. Most of them only need to be instantiated once.